File: apps/windows/SafeWorld.Core.Tests/MatcherTests.cs
using SafeWorld.Core;
using Xunit;

namespace SafeWorld.Core.Tests;

/// <summary>Mirrors packages/core/test/matcher.test.ts — keep the two in sync.</summary>
public class MatcherTests
{
    private static Dictionary<CategoryId, HashSet<string>> Lists(
        IEnumerable<string>? scam = null,
        IEnumerable<string>? gambling = null,
        IEnumerable<string>? adult = null) => new()
    {
        [CategoryId.Scam] = new HashSet<string>(scam ?? Array.Empty<string>()),
        [CategoryId.Gambling] = new HashSet<string>(gambling ?? new[] { "bet365.com" }),
        [CategoryId.Adult] = new HashSet<string>(adult ?? new[] { "pornhub.com" }),
    };

    [Fact]
    public void NormalizeHost_StripsSchemePathPortWwwAndLowercases()
    {
        Assert.Equal("bet365.com", Matcher.NormalizeHost("HTTPS://WWW.Bet365.com:443/path?x=1"));
    }

    [Fact]
    public void NormalizeHost_StripsUserinfoAndTrailingDot()
    {
        Assert.Equal("sub.example.com", Matcher.NormalizeHost("http://user:pass@sub.example.com./a"));
    }

    [Fact]
    public void NormalizeHost_ReturnsEmptyForBlankInput()
    {
        Assert.Equal("", Matcher.NormalizeHost("   "));
    }

    [Fact]
    public void HostMatchesDomain_MatchesExactAndSubdomainsButNotSiblings()
    {
        Assert.True(Matcher.HostMatchesDomain("bet365.com", "bet365.com"));
        Assert.True(Matcher.HostMatchesDomain("m.bet365.com", "bet365.com"));
        Assert.False(Matcher.HostMatchesDomain("notbet365.com", "bet365.com"));
    }

    [Fact]
    public void Decide_BlocksAListedDomainInAnEnabledCategory()
    {
        var d = Matcher.Decide("www.bet365.com", Settings.Defaults(), Lists());
        Assert.True(d.Blocked);
        Assert.Equal("list2", d.Reason);
    }

    [Fact]
    public void Decide_BlocksSubdomainsOfAListedDomain()
    {
        Assert.True(Matcher.Decide("sports.bet365.com", Settings.Defaults(), Lists()).Blocked);
    }

    [Fact]
    public void Decide_DoesNotBlockWhenMasterSwitchIsOff()
    {
        var s = Settings.Defaults();
        s.Enabled = false;
        Assert.False(Matcher.Decide("bet365.com", s, Lists()).Blocked);
    }

    [Fact]
    public void Decide_DoesNotBlockWhenTheCategoryIsDisabled()
    {
        var s = Settings.Defaults();
        s.Categories[CategoryId.Gambling] = false;
        Assert.False(Matcher.Decide("bet365.com", s, Lists()).Blocked);
    }

    [Fact]
    public void Decide_CustomAllowOverridesACategoryBlock()
    {
        var s = Settings.Defaults();
        s.CustomAllow = new List<string> { "bet365.com" };
        Assert.False(Matcher.Decide("bet365.com", s, Lists()).Blocked);
    }

    [Fact]
    public void Decide_CustomBlockBlocksAnOtherwiseAllowedDomain()
    {
        var s = Settings.Defaults();
        s.CustomBlock = new List<string> { "example.com" };
        var d = Matcher.Decide("app.example.com", s, Lists());
        Assert.True(d.Blocked);
        Assert.Equal("custom", d.Reason);
    }

    [Fact]
    public void Decide_AllowsUnlistedDomains()
    {
        Assert.False(Matcher.Decide("wikipedia.org", Settings.Defaults(), Lists()).Blocked);
    }
}

File: apps/ios/SafeWorldCore/Tests/SafeWorldCoreTests/MatcherTests.swift
import XCTest
@testable import SafeWorldCore

/// Mirrors packages/core/test/matcher.test.ts — keep the two in sync.
final class MatcherTests: XCTestCase {
    private func lists(
        scam: [String] = [],
        gambling: [String] = ["bet365.com"],
        adult: [String] = ["pornhub.com"]
    ) -> [CategoryId: Set<String>] {
        [.scam: Set(scam), .gambling: Set(gambling), .adult: Set(adult)]
    }

    // MARK: normalizeHost

    func testNormalizeHostStripsSchemePathPortWwwAndLowercases() {
        XCTAssertEqual(Matcher.normalizeHost("HTTPS://WWW.Bet365.com:443/path?x=1"), "bet365.com")
    }

    func testNormalizeHostStripsUserinfoAndTrailingDot() {
        XCTAssertEqual(Matcher.normalizeHost("http://user:pass@sub.example.com./a"), "sub.example.com")
    }

    func testNormalizeHostReturnsEmptyForBlankInput() {
        XCTAssertEqual(Matcher.normalizeHost("   "), "")
    }

    // MARK: hostMatchesDomain

    func testHostMatchesDomainMatchesExactAndSubdomainsButNotSiblings() {
        XCTAssertTrue(Matcher.hostMatchesDomain("bet365.com", "bet365.com"))
        XCTAssertTrue(Matcher.hostMatchesDomain("m.bet365.com", "bet365.com"))
        XCTAssertFalse(Matcher.hostMatchesDomain("notbet365.com", "bet365.com"))
    }

    // MARK: decide

    func testDecideBlocksAListedDomainInAnEnabledCategory() {
        let d = Matcher.decide(host: "www.bet365.com", settings: .defaults(), blocklists: lists())
        XCTAssertEqual(d, Matcher.BlockDecision(blocked: true, reason: "list2"))
    }

    func testDecideBlocksSubdomainsOfAListedDomain() {
        XCTAssertTrue(Matcher.decide(host: "sports.bet365.com", settings: .defaults(), blocklists: lists()).blocked)
    }

    func testDecideDoesNotBlockWhenMasterSwitchIsOff() {
        var s = Settings.defaults()
        s.enabled = false
        XCTAssertFalse(Matcher.decide(host: "bet365.com", settings: s, blocklists: lists()).blocked)
    }

    func testDecideDoesNotBlockWhenTheCategoryIsDisabled() {
        var s = Settings.defaults()
        s.categories[.gambling] = false
        XCTAssertFalse(Matcher.decide(host: "bet365.com", settings: s, blocklists: lists()).blocked)
    }

    func testDecideCustomAllowOverridesACategoryBlock() {
        var s = Settings.defaults()
        s.customAllow = ["bet365.com"]
        XCTAssertFalse(Matcher.decide(host: "bet365.com", settings: s, blocklists: lists()).blocked)
    }

    func testDecideCustomBlockBlocksAnOtherwiseAllowedDomain() {
        var s = Settings.defaults()
        s.customBlock = ["example.com"]
        XCTAssertEqual(
            Matcher.decide(host: "app.example.com", settings: s, blocklists: lists()),
            Matcher.BlockDecision(blocked: true, reason: "custom")
        )
    }

    func testDecideAllowsUnlistedDomains() {
        XCTAssertFalse(Matcher.decide(host: "wikipedia.org", settings: .defaults(), blocklists: lists()).blocked)
    }
}
